# wgs-varcall

![GitHub Workflow Status (with event)](https://img.shields.io/github/actions/workflow/status/anand-imcm/wgs-varcall/publish.yml)&nbsp;&nbsp;
![GitHub release (with filter)](https://img.shields.io/github/v/release/anand-imcm/wgs-varcall)&nbsp;&nbsp;
[![Open](https://img.shields.io/badge/Open-Dockstore-blue)](https://dockstore.org/workflows/github.com/anand-imcm/wgs-varcall)

> [!TIP]
> To import the workflow into your Terra workspace, click on the above Dockstore badge, and select 'Terra' from the 'Launch with' widget on the Dockstore workflow page.

This repository contains a WDL-based workflow for variant calling and annotation from WGS CRAM files. It uses Google's DeepVariant for genome-wide SNV and indel detection, and Illumina's Gauchian for precise variant calling in the GBA gene region

## Workflow Steps

The workflow is composed of the following steps:

- **GBA Variant Calling**: If `use_gauchian` is `true`, the workflow uses the Illumina [`Gauchian`](https://github.com/Illumina/Gauchian) tool to call variants specifically within the GBA gene.

- **Whole Genome Variant Calling**: If `use_deepvariant` is `true`, the workflow performs whole-genome variant calling using Google [`DeepVariant`](https://github.com/google/deepvariant) to produce a VCF file.

- **Variant Annotation**: When `use_vep` is enabled, the VCF file generated by `DeepVariant` is annotated using the Ensembl [Variant Effect Predictor](https://www.ensembl.org/info/docs/tools/vep/index.html) (`VEP`).

- **Summary**: This step generates an HTML report summarizing the alignment, coverage, and variant call statistics from the previous steps.

## Inputs

### Required

- `main.cram`: [File]
  
- `main.sample_id`: [String]
  
- `main.genome_reference`: [File]

### Optional

- `main.use_gauchian`: [Boolean] If `true`, perform the target variant calling in GBA gene using Gauchian variant caller. Default: `true`.

- `main.use_deepvariant`: [Boolean] If `true`, perform whole-genome variant calling using DeepVariant. Default: `false`.

- `main.use_vep`: [Boolean] If true, annotate DeepVariant's VCF with VEP. Default: `false`.

- `main.vep_cache`: [File] VEP cache in .zip format. This cache is required by the VEP tool for annotation. Use the "merged" Ensembl and RefSeq cache.

- `main.*.cpu`: [Int] Number of CPUs for each task. Default: `32`.

- `main.*.memory_gb`: [Int] Memory in `GB` for each task. Default: `48`.

> [!NOTE]
> A custom `vep_cache.zip` file has been created which contains: [Ensembl GRCh38 release v113 refseq merged cache](https://ftp.ensembl.org/pub/release-113/variation/indexed_vep_cache/homo_sapiens_merged_vep_113_GRCh38.tar.gz) (extracted), [clinvar.vcf.gz](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz) and [clinvar.vcf.gz.tbi](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz.tbi)

## Outputs

The workflow produces the following outputs, organized by the generating tool:

### GBA Variant Calling

- `main.gauchian_tsv`: [File] TSV file containing GBA variant calls from Gauchian.

- `main.gauchian_json`: [File] JSON file containing GBA variant calls and other metadata from Gauchian.

### Whole Genome Variant Calling

- `main.deepvariant_gvcf`: [File] Genome VCF (gVCF) file from DeepVariant containing calls for all sites.

- `main.deepvariant_all_vcf`: [File] VCF file with all variant calls from DeepVariant.

- `main.deepvariant_all_vcf_stats`: [File] Statistics file for the unfiltered DeepVariant VCF.

- `main.deepvariant_filtered_vcf`: [File] A VCF file containing high-confidence variants that have been filtered and normalized. The filtering criteria are `FILTER="PASS"`, `GQ > 20`, `DP > 5`, and `QUAL > 30`. The variant normalization involves splitting the multi-allelic sites and left-aligning indels.

- `main.deepvariant_filtered_vcf_stats`: [File] Statistics file for the filtered DeepVariant VCF.

- `main.deepvariant_summary`: [File] The visual summary report generated by DeepVariant.

### Variant annotation

- `main.vep_annotated_vcf`: [File] VCF file annotated by VEP with variant consequences.

- `main.vep_annotated_tsv`: [File] Annotation table in TSV format derived from `main.vep_annotated_vcf`.

- `main.vep_annotation_stats`: [File] A text file containing summary statistics from the VEP annotation.

### Summary and QC

- `main.qc_report`: [File] A MultiQC report summarizing various QC metrics in HTML format.

- `main.alignment_summary`: [File] An alignment summary of the CRAM file generated by Samtools.

- `main.depth_summary`: [File] Per-contig coverage summary from Mosdepth.

- `main.global_depth_summary`: [File] Global coverage distribution file from Mosdepth.

- `main.qc_plots`: [File] A zip archive of plots generated by MultiQC.

- `main.qc_data`: [File] A zip archive of the data tables used to generate the MultiQC report.

## Components

- [Gauchian v1.0.2](https://github.com/Illumina/Gauchian/releases/tag/v1.0.2) (PolyForm Strict License 1.0.0)

- [DeepVariant v1.9.0]([google/deepvariant](https://github.com/google/deepvariant/releases/tag/v1.9.0)) (BSD-3-Clause license)

- [ensembl-vep release/110.1](https://github.com/Ensembl/ensembl-vep/releases/tag/release%2F110.1) (Apache-2.0 license)

- [samtools v1.22](https://github.com/samtools/samtools) (MIT/Expat License)

- [bcftools v1.22](https://github.com/samtools/bcftools) (MIT/Expat License)

- [mosdepth v0.3.10](https://github.com/brentp/mosdepth) (MIT license)

- [multiqc v1.29](https://github.com/MultiQC/MultiQC) (GPL-3.0 license)
